arangod/Replication2/StateMachines/Document/ActiveTransactionsQueue.cpp

```diff
@@ -0,0 +1,103 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Alexandru Petenchea
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Replication2/StateMachines/Document/ActiveTransactionsQueue.h"
+
+#include <Basics/debugging.h>
+
+#include <algorithm>
+
+namespace arangodb::replication2::replicated_state::document {
+
+/**
+ * Returns the first index which can be released without affecting discarding
+ * any active transactions.
+ */
+LogIndex ActiveTransactionsQueue::getReleaseIndex(LogIndex current) const {
+  if (_logIndices.empty()) {
+    return current;
+  }
+  auto const& idx = _logIndices.front();
+  TRI_ASSERT(idx.second == Status::ACTIVE);
+  return idx.first.saturatedDecrement();
+}
+
+/**
+ * @brief Remove a transaction from the active transactions map and update the
+ * log indices list.
+ * @return true if the transaction was found and removed, false otherwise
+ */
+bool ActiveTransactionsQueue::erase(TransactionId const& tid) {
+  auto it = _transactions.find(tid);
+  if (it == std::end(_transactions)) {
+    return false;
+  }
+
+  auto deactivateIdx =
+      std::lower_bound(std::begin(_logIndices), std::end(_logIndices),
+                       std::make_pair(it->second, Status::ACTIVE));
+  TRI_ASSERT(deactivateIdx != std::end(_logIndices) &&
+             deactivateIdx->first == it->second);
+  deactivateIdx->second = Status::INACTIVE;
+  _transactions.erase(it);
+
+  popInactive();
+
+  return true;
+}
+
+/**
+ * @brief Try to insert an entry into the active transactions map. If the entry
+ * is new, mark the log index as active.
+ */
+void ActiveTransactionsQueue::emplace(TransactionId tid, LogIndex index) {
+  if (_transactions.try_emplace(tid, index).second) {
+    _logIndices.emplace_back(index, Status::ACTIVE);
```

Comment:
Can we add an assertion that a previous empty (in case the queue is not empty) has a lower index?